### PR TITLE
Track Subcontracting Process by Handling Unit

### PR DESCRIPTION
**Pull request by viralkansodiya**

### CODE REVIEW

**Comment by viralkansodiya** [comment]
@agritheory

File: beam/hooks.py [hunk 1]
```diff
@@ -241,3 +246,9 @@
 # auth_hooks = [
 # 	"beam.auth.validate"
 # ]
+
+from erpnext.controllers.subcontracting_controller import SubcontractingController
+
+from beam.beam.overrides.subcontracting_controller import update_stock_ledger
+
+SubcontractingController.update_stock_ledger = update_stock_ledger
```
Comment (agritheory) [inline beam/hooks.py]: Please no overrides inside of hooks.py. We should be using the doctype override in a case like this.

**Review by agritheory — CHANGES_REQUESTED** [review]
We need a test for this also. I would add one to test_handling_unit.py to cover it

File: beam/beam/handling_unit.py [hunk 2]
```diff
@@ -19,7 +19,10 @@ def generate_handling_units(doc, method=None):
 		return doc
 
 	for row in doc.items:
-		if not frappe.get_value("Item", row.item_code, "is_stock_item"):
+		if not (
+			frappe.get_value("Item", row.item_code, "is_stock_item")
+			and frappe.get_value("Item", row.item_code, "enable_handling_unit")
```
Comment (szufisher) [inline beam/beam/handling_unit.py]: get_value can handle multi fields return
frappe.get_value("Item", row.item_code, ["is_stock_item","enable_handling_unit"])

File: beam/beam/overrides/subcontracting_controller.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+						"warehouse": self.supplier_warehouse,
+						"actual_qty": -1 * flt(item.consumed_qty, item.precision("consumed_qty")),
+						"dependant_sle_voucher_detail_no": item.reference_name,
+						"handling_unit": handling_unit,
+					},
+				)
+			)
+
+
+def get_hendling_unit_for_consumption(sle_hu, item):
```
Comment (szufisher) [inline beam/beam/overrides/subcontracting_controller.py]: misspelling?  hendling_unit=>handling_unit?

File: beam/beam/handling_unit.py (hunk 2 above)
Comment (viralkansodiya) [reply beam/beam/handling_unit.py]: Code is Updated

File: beam/beam/overrides/subcontracting_controller.py (hunk 3 above)
Comment (viralkansodiya) [reply beam/beam/overrides/subcontracting_controller.py]: I have fixed it

File: beam/hooks.py (hunk 1 above)
Comment (viralkansodiya) [reply beam/hooks.py]: I have removed monkey patch and added as doctype base class override

**Comment by viralkansodiya** [comment]
@szufisher @agritheory 
please review the latest changes

**Comment by agritheory** [comment]
@viralkansodiya https://open-vsx.org/extension/streetsidesoftware/code-spell-checker has saved me from lots of foolish spelling mistakes

**Comment by viralkansodiya** [comment]
> @viralkansodiya https://open-vsx.org/extension/streetsidesoftware/code-spell-checker has saved me from lots of foolish spelling mistakes

Okay 
Thanks @agritheory

**Comment by agritheory** [comment]
This needs a unit test, I would base it on the [Purchase Invoice one](https://github.com/agritheory/beam/blob/version-14/beam/tests/test_handling_unit.py#L19). Innlude it in the same file after the [manufacture test](https://github.com/agritheory/beam/blob/version-14/beam/tests/test_handling_unit.py#L232). There's a chance we may have to add [pytest-order](https://pypi.org/project/pytest-order/) if it gets weird.

**Review by agritheory — CHANGES_REQUESTED** [review]
This needs a unit test, I would base it on the [Purchase Invoice one](https://github.com/agritheory/beam/blob/version-14/beam/tests/test_handling_unit.py#L19). Include it in the same file after the [manufacture test](https://github.com/agritheory/beam/blob/version-14/beam/tests/test_handling_unit.py#L232). There's a chance we may have to add [pytest-order](https://pypi.org/project/pytest-order/) if it gets weird.

**Comment by viralkansodiya** [comment]
@agritheory 
i have updated the test cases for subcontracting receipt, please review and update me

**Review by agritheory — CHANGES_REQUESTED** [review]
Please rebase, otherwise good to go

**Comment by viralkansodiya** [comment]
@agritheory Please update this PR first, because here I have updated the test cases., so it will help me to resolve Pytest CI in #92

**Comment by viralkansodiya** [comment]
@agritheory 
Now it's ready to merge

**Pull request merged**